cf-custom-resources/lib/dns-cert-validator.js

```diff
@@ -76,76 +76,107 @@ let report = function (
+ * The env hosted zone "env.app.example.com" is created when running "env init" part of the env stack.
+ * Lastly, the function exits until the certificate is validated.
  *
  * @param {string} requestId the CloudFormation request ID
+ * @param {string} appName the name of the application
+ * @param {string} envName the name of the environment
  * @param {string} domainName the Common Name (CN) field for the requested certificate
- * @param {string} hostedZoneId the Route53 Hosted Zone ID
+ * @param {string} subjectAlternativeNames additional FQDNs to be included in the
+ * Subject Alternative Name extension of the requested certificate
+ * @param {string} envHostedZoneId the environment Route53 Hosted Zone ID
+ * @param {string} rootDnsRole the IAM role ARN that can manage domainName
+ * @param {string} isAliasEnabled whether alias is enabled
  * @returns {string} Validated certificate ARN
  */
 const requestCertificate = async function (
   requestId,
+  appName,
+  envName,
   domainName,
   subjectAlternativeNames,
-  hostedZoneId,
+  envHostedZoneId,
+  rootDnsRole,
+  isAliasEnabled,
   region
 ) {
   const crypto = require("crypto");
-  const [acm, route53] = clients(region);
+  const [acm, envRoute53, appRoute53] = clients(region, rootDnsRole);
+  subjectAlternativeNames =
+    isAliasEnabled === "false"
+      ? [`*.${envName}.${appName}.${domainName}`]
+      : subjectAlternativeNames;
   const reqCertResponse = await acm
     .requestCertificate({
-      DomainName: domainName,
+      DomainName: `${envName}.${appName}.${domainName}`,
       SubjectAlternativeNames: subjectAlternativeNames,
       IdempotencyToken: crypto
         .createHash("sha256")
         .update(requestId)
         .digest("hex")
         .substr(0, 32),
       ValidationMethod: "DNS",
+      Tags: [
+        {
+          Key: "copilot-application",
+          Value: appName,
+        },
+        {
+          Key: "copilot-environment",
+          Value: envName,
+        },
+      ],
     })
     .promise();
 
-  let record;
-  for (let attempt = 0; attempt < maxAttempts && !record; attempt++) {
+  let options;
+  let attempt;
+  for (attempt = 0; attempt < maxAttempts; attempt++) {
     const { Certificate } = await acm
       .describeCertificate({
         CertificateArn: reqCertResponse.CertificateArn,
       })
       .promise();
-    const options = Certificate.DomainValidationOptions || [];
-
-    if (options.length > 0 && options[0].ResourceRecord) {
-      record = options[0].ResourceRecord;
-    } else {
-      // Exponential backoff with jitter based on 200ms base
-      // component of backoff fixed to ensure minimum total wait time on
-      // slow targets.
-      const base = Math.pow(2, attempt);
-      await sleep(random() * base * 50 + base * 150);
+    options = Certificate.DomainValidationOptions || [];
+    var areAllResourceRecordsReady = false;
+    for (const option of options) {
```

Comment:
Is it possible that `Certificate.DomainValidationOptions` returns only two options when we expected three options 🤔 and because of we say that all resources are ready when in fact it's not 💭

Reply:
good call! i'll add the validation